src/agent/src/image_rpc.rs

```diff
@@ -44,6 +76,197 @@ impl ImageService {
         oci_path
     }
 
+    fn handle_layer_data(
+        image_data: &ImageData,
+        image_manifest: &mut OciManifest,
+        crypto_config: &CryptoConfig,
+        oci_blob_path: &str,
+    ) -> Result<()> {
+        for layer in image_data.layers.iter() {
+            let layer_digest = layer.clone().sha256_digest();
+
+            if layer.media_type == MEDIA_TYPE_LAYEE_GZIP_ENC
+                || layer.media_type == MEDIA_TYPE_LAYER_ENC
+            {
+                if crypto_config.decrypt_config.is_none() {
+                    break;
+                }
+
+                for layer_desc in image_manifest.layers.iter_mut() {
+                    if layer_desc.digest == layer_digest {
+                        let (layer_decryptor, _dec_digest) = decrypt_layer(
+                            &crypto_config.decrypt_config.as_ref().unwrap(),
+                            layer.data.as_slice(),
+                            layer_desc,
+                            false,
+                        )?;
+                        let mut plaintxt_data: Vec<u8> = Vec::new();
+                        let mut decryptor = layer_decryptor.unwrap();
+
+                        decryptor.read_to_end(&mut plaintxt_data)?;
+                        let layer_name = format!("{:x}", Sha256::digest(&plaintxt_data));
+                        let mut out_file =
+                            File::create(format!("{}/{}", oci_blob_path, layer_name))?;
+                        info!(sl!(), "Saving layer file {}...", layer_name);
+                        out_file.write_all(&plaintxt_data)?;
+                        layer_desc.media_type = manifest::IMAGE_LAYER_GZIP_MEDIA_TYPE.to_string();
+
+                        layer_desc.digest = format!("sha256:{}", layer_name);
+                    }
+                }
+            } else if let Some(layer_name) = layer_digest.strip_prefix("sha256:") {
+                let mut out_file = File::create(format!("{}/{}", oci_blob_path, layer_name))?;
+                info!(sl!(), "Saving layer file {}...", layer_name);
+                out_file.write_all(&layer.data)?;
+            }
+        }
```

Comment:
So what happens for layers that are not `GZIP_ENC` or `ENC` and for which the digest is not a SHA256? I think we should error out in that case because we don't support it.

Reply:
Thanks, I added the error info for non supported digest algo now.